// @markbind/core/src/Site/SiteConfig.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SiteConfig = void 0;
const HEADING_INDEXING_LEVEL_DEFAULT = 3;
/**
 * Represents a read only site config read from the site configuration file,
 * with default values for unspecified properties.
 */
class SiteConfig {
    /**
     * @param siteConfigJson The raw json read from the site configuration file
     * @param cliBaseUrl As read from the --baseUrl option
     */
    constructor(siteConfigJson, cliBaseUrl) {
        this.baseUrl = cliBaseUrl !== undefined
            ? cliBaseUrl
            : (siteConfigJson.baseUrl || '');
        this.enableSearch = siteConfigJson.enableSearch === undefined || siteConfigJson.enableSearch;
        this.faviconPath = siteConfigJson.faviconPath;
        this.headingIndexingLevel = siteConfigJson.headingIndexingLevel || HEADING_INDEXING_LEVEL_DEFAULT;
        this.style = siteConfigJson.style || {};
        this.style.codeTheme = this.style.codeTheme || 'dark';
        this.style.codeLineNumbers = this.style.codeLineNumbers !== undefined
            ? this.style.codeLineNumbers : false;
        this.pages = siteConfigJson.pages || [];
        this.pagesExclude = siteConfigJson.pagesExclude || [];
        this.ignore = siteConfigJson.ignore || [];
        this.externalScripts = siteConfigJson.externalScripts || [];
        this.titlePrefix = siteConfigJson.titlePrefix || '';
        this.titleSuffix = siteConfigJson.titleSuffix || '';
        this.globalOverride = siteConfigJson.globalOverride || {};
        this.timeZone = siteConfigJson.timeZone || 'UTC';
        this.locale = siteConfigJson.locale || 'en-GB';
        this.plugins = siteConfigJson.plugins || [];
        this.pluginsContext = siteConfigJson.pluginsContext || {};
        this.deploy = siteConfigJson.deploy || {};
        this.intrasiteLinkValidation = siteConfigJson.intrasiteLinkValidation || {};
        this.intrasiteLinkValidation.enabled = this.intrasiteLinkValidation.enabled !== false;
        // TODO this should probably be in pluginsContext
        this.plantumlCheck = siteConfigJson.plantumlCheck !== undefined
            ? siteConfigJson.plantumlCheck : true; // check PlantUML's prerequisite by default
    }
}
exports.SiteConfig = SiteConfig;
